### PR TITLE
Dqn paper implementation

### PR DESCRIPTION
**Pull request by edudev**
@AmanDaVinci Can you check that I didn't mess up the hydra stuff with the rebase.
@VaniOFX I remove the use of a smooth L1 loss, and went with just a non-smooth one. Any reason why you used the smooth version?
Also, I left the hidden layer size be 128 (the paper used 512).
I also left Adam as the default optimizer, though this can easily be changed in the config.
I also left some comments on where I didn't follow the paper or couldn't.

### CODE REVIEW

File: dqn.py [hunk 1]
```diff
@@ -99,6 +115,9 @@ def compute_q_vals(self, states, actions):
 
         return result
 
+    def clear_max_q_val(self):
+        self.max_q_val = torch.tensor(0.0).to(self.device)
```
Comment (VaniOFX) [inline dqn.py:119]: Is there any reason why we might need something more than just a simple float?

File: dqn.py [hunk 2]
```diff
@@ -128,14 +153,15 @@ def train(self, batch_size, sample_memory=False):
...
-        rewards[rewards > 1.0] = 1.0
-        rewards[rewards < -1.0] = -1.0
+        rewards = torch.clamp(rewards, -1.0, 1.0)
 
         q_val = self.compute_q_vals(states, actions)
         with torch.no_grad():  # Don't compute gradient info for the target (semi-gradient)
             target = self.compute_targets(batch_size, rewards, next_states, dones)
 
-        loss = F.smooth_l1_loss(q_val, target)
+        # loss = F.smooth_l1_loss(q_val, target)
```
Comment (VaniOFX) [inline dqn.py:162]: I put the smooth version as that was suggested in the lab by the TAs. However, in the paper they use MSE loss which is L2 loss rather than L1.

File: dqn.py [hunk 3]
```diff
@@ -202,28 +234,40 @@ def main(config: RLConfig) -> None:
...
             state = next_state
             steps += 1
 
             if done:
                 average_steps.append(steps)
-                if episode % 20 == 0:
+                rewards.append(episode_reward)
+
+                if (episode + 1) % 20 == 0:
                 # if episode % 1 == 0:
```
Comment (VaniOFX) [inline dqn.py]: Maybe also delete the unnecessary commeted code.

File: dqn.py [hunk 4]
```diff
@@ -202,28 +234,40 @@ def main(config: RLConfig) -> None:
...
-                    log.info(f"Episode: {episode:05d}/{config.num_episodes:05d}\t\t "
-                             f"Score: {np.mean(average_steps):.1f}\t\t "
-                             f"Max-|Q|: {agent.get_max_q_val():.1f}\t\t "
+                    log.info(f"Episode: {episode + 1:5d}/{config.num_episodes:5d}\t\t "
+                             f"#Steps: {np.mean(average_steps):7.1f}\t\t "
+                             f"Reward: {np.sum(rewards):7.1f}\t\t "
```
Comment (VaniOFX) [inline dqn.py]: Why sum all the episode rewards rather than mean?

File: dqn.py (hunk 1 above)
Comment (edudev) [reply dqn.py:119]: Perhaps so that when you take the max, there is no data transfer between CPU and GPU in this line:
```
        self.max_q_val = torch.max(self.max_q_val, result.abs().max())
```
I just like to keep all tensors on the same device.

I guess it absolutely doesn't matter for our task?

File: dqn.py (hunk 2 above)
Comment (edudev) [reply dqn.py:162]: You're right! I read this line incorrectly.
> Because the absolute value loss function `|x|` has a derivative of -1 for all negative values of `x` and a derivative of 1 for all positive values of `x`, clipping the squared error to be between -1 and 1 corresponds to using an absolute value loss function for errors outside of the (-1,1) interval.

I will change it to being MSE + clipping.

File: dqn.py (hunk 4 above)
Comment (edudev) [reply dqn.py]: Leftover from previous code, when I didn't calculate a per-episode total reward, but accumulated all rewards in a single list.
I'll correct that.

File: dqn.py (hunk 3 above)
Comment (edudev) [reply dqn.py]: Sure can do, though I still sometimes reduce the printing to every episode, so that I can debug faster.
Maybe I'll make it a param.

File: dqn.py [hunk 5]
```diff
@@ -58,19 +64,29 @@ def forward(self, input):
...
-        self.optimizer = Adam(self.Q_model.parameters(), lr=learning_rate)
+        self.optimizer = self.build_optimizer(optimizer, learning_rate)
         self.max_q_val = torch.tensor(0.0).to(self.device)
         self.use_target_net = use_target_net
         self.update_target_freq = update_target_freq
 
+        self.clear_max_q_val()
+
+    def build_optimizer(self, name, learning_rate):
+        # perhaps not the best way, but gives good flexibility
```
Comment (AmanDaVinci) [inline dqn.py:85]: yes, seems simple and reasonable for now.

**Review by AmanDaVinci — APPROVED** [review]
Should work fine with the configuration system.

**Review by VaniOFX — APPROVED** [review]
